flink/v1.19/flink/src/main/java/org/apache/iceberg/flink/maintenance/operator/MonitorSource.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.maintenance.operator;
+
+import java.io.IOException;
+import java.util.Iterator;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.connector.source.Boundedness;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.api.connector.source.SourceReaderContext;
+import org.apache.flink.api.connector.source.util.ratelimit.RateLimitedSourceReader;
+import org.apache.flink.api.connector.source.util.ratelimit.RateLimiter;
+import org.apache.flink.api.connector.source.util.ratelimit.RateLimiterStrategy;
+import org.apache.flink.core.io.SimpleVersionedSerializer;
+import org.apache.flink.core.memory.DataInputDeserializer;
+import org.apache.flink.core.memory.DataOutputSerializer;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.base.MoreObjects;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** Monitors an Iceberg table for changes */
+public class MonitorSource extends SingleThreadedIteratorSource<TableChange> {
+  private static final Logger LOG = LoggerFactory.getLogger(MonitorSource.class);
+
+  private final TableLoader tableLoader;
+  private final RateLimiterStrategy rateLimiterStrategy;
+  private final long maxReadBack;
+
+  /**
+   * Creates a {@link org.apache.flink.api.connector.source.Source} which monitors an Iceberg table
+   * for changes.
+   *
+   * @param tableLoader used for accessing the table
+   * @param rateLimiterStrategy limits the frequency the table is checked
+   * @param maxReadBack sets the number of snapshots read before stopping change collection
+   */
+  public MonitorSource(
```

Comment:
I found it a little strange with read limiter and iterator here.

Let me bring up a couple of alternatives to discuss

1. MonitorSource enumerator can emit `TableChangeSplit` and reader just extract the `TableChange` and emit to downstream Trigger operator
2. MonitorSource enumerator just emit `Snapshot` split and reader can process snapshot and construct `TableChange` from manifest info.

I like the second option better, the split id could be table UUID + snapshot Id.

Reply:
We want to have a single TableChange event on start/restart, as we don't want to double trigger Maintenance Tasks on startup. If we read parallel, I don't see how can we archive this.
If we have so much data, that parallel readers are needed, we probably just want to trigger the Maintenance Tasks and start from scratch.

Realistically you want to trigger an incremental compaction after a few 10s of commits, a few 100MBs of data, or after a specific timeout.

I don't think you want to calculate the accumulated value of a 1000 commits to decide what to do.

Reply:
Here are the steps if I understood correctly

1. the MonitorSource does some table scan and calculation
2. emit a TableChange record to the trigger operator, 
3. trigger does the table scan again to decide if the actual maintenance task/action

I feel we can just combine them into one source operator. don't see much value of two separate operators here.

code sharing maybe can be achieved via util class (e.g. on deciding what to do).

Reply:
The `TriggerManager` doesn't do any scanning. It's only using the `TableChange` events to decide if the Maintenance Tasks are need to be started, or not. After the decision, we create an object with the current snapshot of the table, and with a current timestamp (this will be the actual Trigger), but we still don't do any table scanning (just read the current snapshot of the table). Only the actual Maintenance Tasks will read/scan the tables, but this only happens when they are triggered, and they are usually done in a way, which is very different than the `MonitorSource` scans.
Also we would like to reuse the `TriggerManager` logic in the PostCommitTopology and when the maintenance is done in a separate job. We can separate out the `TriggerManager` to just a java code, but I feel that using the Flink provided operator abstraction is a nice way to do it, especially since we can make sure, that the operators are chained.